File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        curr= root
        stack = []
        res = []
        while stack or curr:
            print(curr.val)
            if stack:
                curr = stack.pop()
            if curr.right:
                stack.append(curr.right)
            while curr.left != None:
                stack.append(curr.left)
                left = curr.left
                curr.left = None
                curr = left
            print([(da.val, da.left) for da in stack])
            res.append(curr.val)
            if len(res) == k:
                return res[-1]
            
        return -1

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        stack<pair<TreeNode*, int>> st;
        int sol{};
        st.push({root, 1 << 31});
        while (!st.empty()) {
            auto [node, maxi] = st.top();
            st.pop();
            if (!node) continue;
            if (node->val >= maxi) {
                sol++;
                std::cout << node->val << std::endl;
            }
            st.push({node->left, std::max(maxi, node->val)});
            st.push({node->right, std::max(maxi, node->val)});
        }


        return sol;
    }
};
